Implement redux

Haywayaheadshot opened the pull request:
## What?
   - I implemented redux in the project. I refactored the code to improve optimization and do away with the API folder, and API calls in components.

## Why?

   - To ensure the application has a single source of truth.

## How?
   - I added redux toolkit for state management
   - I converted API calls to thunks in slices
   - I received all queries in NewArrivals.js through the redux store to ensure a single source of truth
   - I deleted the initial API calls in NewArrivals.js
   - I created store to load filter categories
   - I created actions and reducers for filtering products on display.
   - I created actions and reducers to read an individual products and list related items to that product.

Haywayaheadshot committed:
install redux and toolkit

Haywayaheadshot committed:
wrap routes in Provider and supply store

Haywayaheadshot committed:
receive new arrivals action

Haywayaheadshot committed:
configure store

Haywayaheadshot committed:
create actions and slice for new products

Haywayaheadshot committed:
fetch all queries for loading products from store

Haywayaheadshot committed:
initialize store and its product selecting vars

Haywayaheadshot committed:
move new arrivals to new folder

Haywayaheadshot committed:
create redux for loading by price

Haywayaheadshot committed:
create redux for loading by sell

Haywayaheadshot committed:
move newArrivals action to a diff module

Haywayaheadshot committed:
add load all categories to store

Haywayaheadshot committed:
receive categories from store and del inApp call

Haywayaheadshot committed:
create redux for getting all categories

Haywayaheadshot committed:
optimize component to increase performance

Haywayaheadshot committed:
receive fetchedCategories with deconstructuring

Haywayaheadshot committed:
compile all actions and reducers of products

Haywayaheadshot committed:
delete filteredResults API call

Haywayaheadshot committed:
refactor loadFilteredResults to dispatch action

Haywayaheadshot committed:
add filteredProductsReducer to store

Haywayaheadshot committed:
add fetchFilteredProducts api

Haywayaheadshot committed:
create actions and reducers for filteredProducts

Haywayaheadshot committed:
format document

Haywayaheadshot committed:
receive departments from redux

Haywayaheadshot committed:
comment out getDepartments

Haywayaheadshot committed:
receive and map departments from redux

Haywayaheadshot committed:
move api file to redux folder

Haywayaheadshot committed:
add departments reducer

Haywayaheadshot committed:
add actions and reducer for department

Haywayaheadshot committed:
add updateFilteredProducts logic

Haywayaheadshot committed:
change api import location

Haywayaheadshot committed:
consolidate products end point to adhere to DRY

Haywayaheadshot committed:
import products api to include required endpoints

Haywayaheadshot committed:
delete getDepartments and listProducts

Haywayaheadshot committed:
call listProducts from redux

Haywayaheadshot committed:
call listRelated and readProduct from redux

Haywayaheadshot committed:
include endpoint for a product

Haywayaheadshot committed:
add reducer to load a product

Haywayaheadshot committed:
add action to list products

Haywayaheadshot committed:
add listProducts reducer cases

Haywayaheadshot committed:
create action to list products

Haywayaheadshot committed:
add list Products reducer

Haywayaheadshot committed:
add action and reducer for listRelated

Haywayaheadshot committed:
create actions and reducer for transactions

Haywayaheadshot committed:
comment out getTransactions API call

Haywayaheadshot committed:
receive getTransactions action

Haywayaheadshot committed:
delete stray console.log

Haywayaheadshot committed:
include transactionsEndPointApi

Haywayaheadshot committed:
comment initiateTransaction API call

Haywayaheadshot committed:
call getTransactionFees and InitiateTransaction

Haywayaheadshot committed:
and transactionEndPointApi

Haywayaheadshot committed:
add transaction actions, slice and reducer

Haywayaheadshot committed:
created redux for all user APIs

Haywayaheadshot committed:
delete old initiateTransaction logic

Haywayaheadshot committed:
add ordersEndPoint Api

Haywayaheadshot committed:
add orders reducer

Haywayaheadshot committed:
create actions and reducer for orders

Haywayaheadshot committed:
modify how data is sent to flutterwave

Haywayaheadshot committed:
delete userAPi module import

DavidQuartz commented:
Let's Implement the vendor APIs after this PR has been reviewed and merged. The commits are pretty long now @Haywayaheadshot

Haywayaheadshot committed:
create axios instance

Haywayaheadshot committed:
put actions in try-catch block

Haywayaheadshot committed:
include initial state to slices

DavidQuartz commented:
> @DavidQuartz I have implemented all but the axios request.use and request.set configurations. I am still a bit confused as to how to do that so I am reading more on it. Can I create a separate issue for that for feature reference?

Yes defintely. I will review this PR and merge if everything looks good. Thank you @Haywayaheadshot

Haywayaheadshot commented:
@DavidQuartz You're welcome. I just realised I still have to implement the vendor API. Working on it now

Haywayaheadshot committed:
comment old vendor api call

Haywayaheadshot committed:
call creatingCategory action and destructure error

Haywayaheadshot committed:
include vendor category and product endpoints

Haywayaheadshot committed:
include vendor reducer

Haywayaheadshot committed:
create vendor actions and reducer

Haywayaheadshot committed:
delete vendor API module

Haywayaheadshot committed:
destructing success and error of creating products

Haywayaheadshot committed:
rename success and error variables

Haywayaheadshot commented:
@DavidQuartz Please commence the review

DavidQuartz commented on Qfront/src/redux/vendor/action.js:
```diff
@@ -0,0 +1,34 @@
+        });
+        return response.data
+    } catch (error) {
+        rejectWithValue(error);
```
you should always return the `rejectWithValue` otherwise nothing gets returned

DavidQuartz commented on Qfront/src/redux/api.js:
```diff
@@ -0,0 +1,18 @@
+    baseURL: process.env.REACT_APP_API_URL,
+});
+
+export const baseUrl = axiosInstance;
```
`baseUrl` is not a descriptive name for the axios instance. `axiosInstance` is much better.

DavidQuartz commented on Qfront/src/redux/api.js:
```diff
@@ -0,0 +1,18 @@
+import axios from 'axios';
+
+// Create base url for axios to
+const axiosInstance = axios.create({
```
You can export this `const` as a named export at the end of the file, instead of reassigning it to another variable and export that

DavidQuartz commented on Qfront/src/redux/departments/Departments.js:
```diff
@@ -0,0 +1,33 @@
+        });
+        builder.addCase(getAllDepartments.rejected, (state, action) => {
+            state.success = false;
+            state.departmentsApiError = action.payload;
```
What is the structure of the error? We only need the error message, and not the entire error object

DavidQuartz commented on Qfront/src/redux/orders/Orders.js:
```diff
@@ -0,0 +1,47 @@
+        builder.addCase(createGuestOrder.rejected, (state, action) => {
+            state.loading = false;
+            state.success = false;
+            state.guestDataError = action.payload;
```
We only need the relevant field out of the entire error payload

DavidQuartz commented on Qfront/src/redux/orders/Orders.js:
```diff
@@ -0,0 +1,47 @@
+            state.guestDataError = action.payload;
+        });
+        builder.addCase(createGuestOrder.pending, (state) => {
+            state.loading = true;
```
In pending, we should set error and success states also to their default states

DavidQuartz commented on Qfront/src/redux/orders/Orders.js:
```diff
@@ -0,0 +1,47 @@
+            state.loading = false;
+            state.authData = action.payload;
+        });
+        builder.addCase(createAuthOrder.pending, (state) => {
```
In pending, we should set error and success states also to their default states. Please make this change in all other reducers necessary

DavidQuartz commented on Qfront/src/redux/orders/Orders.js:
```diff
@@ -0,0 +1,47 @@
+    initialState,
+    reducers: {},
+    extraReducers: (builder) => {
+        builder.addCase(createAuthOrder.fulfilled, (state, action) => {
```
In fulfilled, we should set error state to their default state. Please make this change in all other reducers necessary

DavidQuartz commented on Qfront/src/redux/orders/Orders.js:
```diff
@@ -0,0 +1,47 @@
+            state.success = false;
+            state.authDataError = action.payload;
+        });
+        builder.addCase(createGuestOrder.fulfilled, (state, action) => {
```
In fulfilled, we should set error state to their default state

DavidQuartz commented on Qfront/src/redux/product/filteredProducts/actions.js:
```diff
@@ -0,0 +1,62 @@
+// create action to list products
+export const listProducts = createAsyncThunk("loadProducts/listProducts", async( { params }, { rejectWithValue } ) => {
+  try {
+    const query = {
```
Is params always supplied to this action?
I see nothing passed [here](https://github.com/Oniocha/Qargo-Ecommerce/pull/11/files#diff-d70b1568fcc78bad0e342846f883db2a81f86af4529b3997a34968452c299d37R26)

In that case the params would be undefined and the request will be made with it. Only construct the query object if params is present.

DavidQuartz commented on Qfront/src/redux/product/loadProducts/actions.js:
```diff
@@ -0,0 +1,57 @@
+// Create action to load by price
+export const getProductsByPrice = createAsyncThunk("LOAD_BY_PRICE", async( _, { rejectWithValue }) => {
+    try {
+       const getProductsByPriceApi = `${productsEndPointApi}?sortBy=price&order=asc&limit=10`;
```
We should avoid hardcoding moving forward. We can have these as default values for the thunk's parameters rather than hard coding.

DavidQuartz commented on Qfront/src/redux/product/loadProducts/actions.js:
```diff
@@ -0,0 +1,57 @@
+       const response = await baseUrl.get(getProductsByPriceApi);
+       return response.data.data;
+    } catch (error) {
+        rejectWithValue(error)
```
You forgot the return keyword. Please make this change wherever necessary

DavidQuartz commented on Qfront/src/redux/product/loadProducts/actions.js:
```diff
@@ -0,0 +1,57 @@
+// create fetch action for new Arrivals
+export const getNewArrivals = createAsyncThunk("FETCH_PRODUCTS", async ( _, { rejectWithValue }) => {
+    try {
+        const getNewArrivalsApi = `${productsEndPointApi}?sortBy=sold&order=desc&limit=10`;
```
We should avoid hardcoding moving forward. We can have these as default values for the thunk's parameters rather than hard coding.

DavidQuartz commented on Qfront/src/Pages/CheckoutPage/CheckOutPage.jsx:
```diff
@@ -105,6 +102,7 @@ const CheckOutPage = () => {
 
   useEffect(() => {
     handleMobileSelector();
+    dispatch(getTransactionFees({cost: 14}))
```
Why do we have 14 hardcoded?

DavidQuartz commented on Qfront/src/Pages/Shop/Shop.jsx:
```diff
@@ -2,67 +2,38 @@ import React, { useState, useEffect } from "react";
-        setSkip(0);
-      }
-    });
+    dispatch(getFilteredProducts(skip, limit, newFilters))
```
Did you test this? Passing 3 parameters to the action? The `options` parameter is only the first one

DavidQuartz commented on Qfront/src/Pages/Shop/Shop.jsx:
```diff
@@ -2,67 +2,38 @@ import React, { useState, useEffect } from "react";
-    });
+    const filter = myFilters.filters
+    // This function might need more testing
+    dispatch(updateFilteredProducts(toSkip, limit, filter))
```
Did you test this? Passing 3 parameters to the action? The `options` parameter is only the first one

DavidQuartz commented on Qfront/src/Pages/Shop/Shop.jsx:
```diff
@@ -78,11 +49,11 @@ const Shop = () => {
   useEffect(() => {
-    init();
+    dispatch(getAllCategories())
     loadFilteredResults(limit, skip, myFilters.filters);
```
Did you test this? Passing 3 parameters to the action? The `options` parameter is only the first one

DavidQuartz reviewed (changes requested):
Good work @Haywayaheadshot. Kindly checkout the comments I added.

Haywayaheadshot committed:
return reject with error and change baseUrl export to axiosInstance

Haywayaheadshot committed:
Return error message, instead of error object

DavidQuartz committed:
Made some fixes

Haywayaheadshot committed:
set states to default state
- In builder cases unchanged variables are set to their default states

Haywayaheadshot committed:
Corrections
- Del hardcoded cost and properly dispatch data to getFilteredProducts
- and updateFilteredProducts

Haywayaheadshot commented:
@DavidQuartz I have made the corrections we spoke about yesterday

DavidQuartz commented on Qfront/src/Pages/Admin/AddCategory.jsx:
```diff
@@ -26,17 +29,21 @@ const AddCategory = () => {
-      }
-    });
+    const access = token;
+    dispatch(createCategory(user._id, access, { name }))
```
this thunk is receiving it's parameters wrongly. Please check to make sure all the thunks are receiving parameters as one argument (could be an object), not as multiple arguments

Haywayaheadshot committed:
Dispatch data for
- createCategory. Which ended succesful
- createProduct. Which had a issues sending formData

Haywayaheadshot committed:
return page logic to initial state

Haywayaheadshot committed:
include more logic to support dispatch

DavidQuartz committed:
Added logger

DavidQuartz committed:
REmoved serializable check

DavidQuartz committed:
fixed params

DavidQuartz committed:
fixed user redux bugs

DavidQuartz committed:
Fixed vendor issues

DavidQuartz committed:
Fixed transactions
